File: Algorithms/cpp/insertion_sort.cpp
/**
========================================================================
@file insertion_sort.cpp

Big O:
    worst case: O(n^2)
    best case:  O(n)

: zach wolpe
: zach.wolpe@medibio.com.au
: 14 July 2023
========================================================================
*/



#include <iostream>
using namespace std;

void insertionSort(int array[], int size) {
    for (int i=1; i<size; i++) {
        int temp = array[i];
        int j = i-1;
        // order of operations!
        while (j > -1 && temp < array[j]) {
            array[j+1] = array[j];
            array[j] = temp;
            j--;   
        }
    };
}


int main() {
    int myArray[] = {6,4,2,5,3,4,2};
    int size = sizeof(myArray) / sizeof(myArray[0]); // number of bytes in array
    insertionSort(myArray, size);
    for (auto value : myArray) {
        cout << value << " ";
    }
    return 0;
};

File: Algorithms/python/insertion_sort.py
"""
========================================================================
@file insertion_sort.py

Big O:
    worst case: O(n^2)
    best case:  O(n)

: zach wolpe
: zach.wolpe@medibio.com.au
: 18 July 2023
========================================================================
"""

def insertion_sort(array):
    for i in range(1, len(array)):
        temp = array[i]
        j = i-1
        while j > -1 and temp < array[j]:
            array[j+1] = array[j]
            array[j] = temp
            j -= 1
    return array


if __name__ == '__main__':
    
    array = [1, 222, 233, 4, 5, 6, 97, 8, 9]
    array = insertion_sort(array)
    [print(i, end=' -> ') for i in array]
    print()